flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -1263,11 +1263,11 @@ public void start() {
         public void onProcessingTime(long timestamp) {
             if (nextWatermark != Long.MIN_VALUE) {
                 long globalWatermark = lastGlobalWatermark;
-                // TODO: refresh watermark while idle
                 if (!(isIdle && nextWatermark == propagatedLocalWatermark)) {
                     globalWatermark = watermarkTracker.updateWatermark(nextWatermark);
                     propagatedLocalWatermark = nextWatermark;
                 } else {
+                    globalWatermark = watermarkTracker.getWatermark();
                     LOG.info(
                             "WatermarkSyncCallback subtask: {} is idle",
                             indexOfThisConsumerSubtask);
```

Comment:
can you log `lastGlobalWatermark`, `globalWatermark` and `isIdle` values here.

Reply:
Also, how about we add more logs at [1] ? Primarily, important state values of  `lastEmittedRecordWatermark`  and `lastWatermark` 

Also, would be great if we can publish a metric at [1] the first time we set `isIdle` to True
 

```
for (Map.Entry<Integer, ShardWatermarkState> e : shardWatermarks.entrySet()) {
            Watermark w = e.getValue().lastEmittedRecordWatermark;
            // consider only active shards, or those that would advance the watermark
            if (w != null
                    && (e.getValue().lastUpdated >= idleTime
                            || e.getValue().emitQueue.getSize() > 0
                            || w.getTimestamp() > lastWatermark)) {
``` 

[1]  https://github.com/lyft/flink/blob/d95cf630c628acbfb699fb63a765f95cb8139ca1/flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java#L1183

Reply:
> can you log `lastGlobalWatermark`, `globalWatermark` and `isIdle` values here.
Check out ln 1279. That logs local, global, delta, timeouts, queues, and queue sizes. Logging here would be redundant I believe. I will add isIdle to that log though.

Reply:
My idea is to add an alert based on the metric to allow the team to quickly go through the logs.

Reply:
agreed. I've added an isIdle metric in the PeriodicWatermarkEmitter and we can write the prom query to detect a 0 to 1 change